sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,396 @@
+    >>> vd = ValueDropper(missing_distribution=missing_distribution,
+    ...                   random_state=0)
+    >>> vd.transform(X, y)
+    array([[  0.,   1.,   2.],
+           [  3.,   4.,   5.],
+           [  6.,   7.,   8.],
+           [  9.,   0.,   1.],
+           [  2.,   3.,   4.],
+           [ nan,   9.,   8.],
+           [ nan,  nan,   5.],
+           [ nan,   8.,   9.],
+           [ nan,   4.,   3.],
+           [  2.,   1.,   1.],
+           [ nan,   2.,   3.]])
+    >>> # MCAR missingness
+    >>> vd = ValueDropper(missing_distribution=abs_missing_rate,
+    ...                   random_state=0)
+    >>> vd.transform(X, y)
+    array([[  0.,   1.,   2.],
+           [  3.,   4.,   5.],
+           [  6.,   7.,  nan],
+           [  9.,   0.,   1.],
+           [ nan,   3.,   4.],
+           [  8.,   9.,   8.],
+           [  1.,   0.,  nan],
+           [  7.,   8.,   9.],
+           [  5.,   4.,   3.],
+           [ nan,  nan,   1.],
+           [  1.,  nan,   3.]])
+    >>> # Upscale the missing_distribution to add more missing values
+    >>> # Explicitly set copy=False for inplace dropping of values
+    >>> vd = ValueDropper(missing_distribution=2 * abs_missing_rate,
+    ...                   copy=False, random_state=0)
+    >>> _ = vd.transform(X, y)
+    >>> X
+    array([[  0.,   1.,   2.],
+           [  3.,   4.,   5.],
+           [ nan,   7.,  nan],
+           [  9.,  nan,   1.],
+           [ nan,   3.,   4.],
+           [  8.,  nan,   8.],
+           [  1.,   0.,  nan],
+           [  7.,   8.,  nan],
+           [  5.,   4.,   3.],
+           [ nan,  nan,  nan],
+           [ nan,  nan,   3.]])
+    """
+
+    def __init__(self, missing_values="NaN",
+                 missing_distribution=None, copy=True, random_state=None):
+        self.missing_values = missing_values
+        self.missing_distribution = missing_distribution
+        self.copy = copy
+        self.random_state = random_state
+
+    def transform(self, X, y=None):
+        """Drop values from ``X`` according to the given distribution.
+
+        Parameters
+        ----------
+
+        X : ndarray like of shape (n_features, n_samples)
+            Data, in which the values must be dropped and set to
+            ``missing_values``.
+
+        y : array-like, shape = (n_samples,), optional for MCAR
+            Target relative to X for classification or regression;
+            When missing_distribution is not a dict (for MCAR missingness),
+            ``y`` need not be passed.
+        """
+        # Validate missing_values and generate missing_mask
+        if ((isinstance(self.missing_values, str) and
+                (self.missing_values.lower() == "nan")) or
+                np.isnan(self.missing_values)):
+            missing_values = np.nan
+            missing_mask = np.isnan(X)
+        else:
+            missing_values = self.missing_values
+            missing_mask = X == self.missing_values
```

Comment:
For me X must be without nan (or equivalent for int). I think it will simplify the code a lot to remove definitely missing_mask.